### ISSUE TITLE
~10-13 GiB of extra memory usage observed when post-training Qwen3-30B-A3B on 1x DGXH100

### ISSUE THREAD

**Issue by rao-ashish** [verl-project/verl#6366]
### System Info

```
----------Python Info----------
Version      : 3.12.13
Compiler     : GCC 11.4.0
Build        : ('main', 'Mar  4 2026 09:23:07')
Arch         : ('64bit', 'ELF')
------------Pip Info-----------
Version      : 26.0.1
Directory    : /usr/local/lib/python3.12/dist-packages/pip
vllm         : 0.17.0
sglang       : not found.
ray          : 2.54.0
torch        : 2.10.0+cu129
----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /opt/workspace/RL/verl/verl
Commit Hash  : 460ccf3c2f19ae537f7de53243c89df69d746f05
----------Platform Info----------
Platform     : Linux-5.15.0-1055-aws-x86_64-with-glibc2.35
system       : Linux
node         : pool0-0038
release      : 5.15.0-1055-aws
version      : #60~20.04.1-Ubuntu SMP Thu Feb 22 15:49:52 UTC 2024
----------Environment----------
CUDA Runtime : 12.9
CUDA Compiler : Cuda compilation tools, release 12.9, V12.9.86
----------System Info----------
CPU Memory      : 1999.96 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA H100 80GB HBM3
GPU 1   Memory  : 79.65 GB
GPU 2   Type    : NVIDIA H100 80GB HBM3
GPU 2   Memory  : 79.65 GB
GPU 3   Type    : NVIDIA H100 80GB HBM3
GPU 3   Memory  : 79.65 GB
GPU 4   Type    : NVIDIA H100 80GB HBM3
GPU 4   Memory  : 79.65 GB
GPU 5   Type    : NVIDIA H100 80GB HBM3
GPU 5   Memory  : 79.65 GB
GPU 6   Type    : NVIDIA H100 80GB HBM3
GPU 6   Memory  : 79.65 GB
GPU 7   Type    : NVIDIA H100 80GB HBM3
GPU 7   Memory  : 79.65 GB
GPU 8   Type    : NVIDIA H100 80GB HBM3
GPU 8   Memory  : 79.65 GB
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

- Job launch script: https://gist.github.com/rao-ashish/8ef9c953196c2dd8fc2cf2bc39e77c48
- Verl commit: 460ccf3c2f19ae537f7de53243c89df69d746f05
- Container: verlai/verl:vllm017.latest
- vLLM version: 0.17.0
- Megatron version: 0.16.0

### Expected behavior

~10-13 GiB of excess memory usage is observed when post-training Qwen3-30B-A3B on GSM8K with 1x DGXH100 node (8x H100-SXM) and when using the Megatron + vLLM backends (job launch script / config is attached above). Setting `enforce_eager=True` still results in ~10 GiB of excess memory usage.

In [this](https://github.com/rao-ashish/verl/tree/409d857574fa7fe3ce68b53b3a90a169d7e63a3c) fork of VeRL, I added some extra logging to show actual memory usage, and I consistently see ~13 GB of excess usage. For example, the log statement below indicates that 13.32 GiB is in-use after `sleep_replicas` is complete. The config turns on host offloading for actors and inference engines, so we expect ~0 GiB memory usage at this point (both train state and inference state should be offloaded to the host). The extra logging added shows that ~7.22 GiB is still taken up at this point by vLLM, and ~5.45 GB is taken up by the ActorRolloutRef worker. This 13 GiB overhead stays consistent across all training steps / stages.

```
(WorkerDict pid=4019404) [0m fit: after sleep_replicas (gen) (per-process GPU memory, GB): device used/total: 13.32/79.65, n_procs: 2
(WorkerDict pid=4019404) [0m   pid=4024990 mem=    7.22 GB name=VLLM::Worker_TP
(WorkerDict pid=4019404) [0m   pid=4019404 mem=    5.45 GB name=ray::WorkerDict <-- self
```

As mentioned above, setting `enforce_eager=True` reduces the memory overhead to ~10 GiB.

This memory overhead forces us to use small microbatch sizes (larger microbatch sizes were observed to OOM), which reduces MFU.

**Comment by Kirrito-k423** [comment]
Thank you for reporting this memory overhead issue.

**Analysis**:

The 10-13 GiB extra memory usage on a single H100 during post-training is significant. This could be caused by several factors:

1. **Non-optimal tensor layout**: Qwen3 MoE models have complex tensor partitioning. If weight loading or gradient communication creates intermediate tensors that are not properly sharded, memory can balloon.

2. **Gradient accumulation buffers**: With , the trainer may be allocating larger intermediate buffers than necessary for your token budget (max_prompt_length + max_response_length).

3. **vLLM vs Megatron weight sync**: If you are using vLLM for rollout, the IPC-based weight transfer can create temporary tensor copies that persist in GPU memory.

**Diagnostic steps**:

1. Could you share your exact training configuration? Specifically: , , and  settings.

2. Please run  during training to capture the peak memory footprint and when it occurs (which training phase: rollout, old_log_prob, update_actor, etc.).

3. Check if the extra memory appears consistently at the same training step or varies across steps.

Would you like to share the relevant portions of your training script?
